### PR TITLE
addCustomizedSubstatusCodeForClientGenerated500

### PR DESCRIPTION
**Pull request by xinlian12**
**Issue:**
For certain situations, the request may fail with 500, however the current diagnostics lacking details to identify what could have caused 500

Example Diagnostics:
```
 "responseStatisticsList": [
                    {
                        "storeResult": {
                            "storePhysicalAddress": "rntbd://membership-prod-prod-cosmos-eastus2.documents.azure.com:36278/apps/7dcbfec6-2f89-4b40-9a14-c692e4b05ed4/services/d46ffeae-93db-45b2-a2ef-765415eea39b/partitions/f7af9b8a-9ce4-4af6-85d3-fb7ab7c28d50/replicas/133112526105274695s/",
                            "lsn": -1,
                            "globalCommittedLsn": -1,
                            "partitionKeyRangeId": null,
                            "isValid": false,
                            "statusCode": 500,
                            "subStatusCode": 0,
                            "isGone": false,
                            "isNotFound": false,
                            "isInvalidPartition": false,
                            "isThroughputControlRequestRateTooLarge": false,
                            "requestCharge": 0.0,
                            "itemLSN": -1,
                            "sessionToken": null,
                            "backendLatencyInMs": null,
                            "retryAfterInMs": null,
                            "exceptionMessage": "Unknown server error occurred when processing this request. If the issue persists, please contact Azure Support: http://aka.ms/azure-support",
                            "exceptionResponseHeaders": "{}",
                            "replicaStatusList": [],
                            "transportRequestTimeline": null,
                            "rntbdRequestLengthInBytes": 0,
                            "rntbdResponseLengthInBytes": 0,
                            "requestPayloadLengthInBytes": 0,
                            "responsePayloadLengthInBytes": 0,
                            "channelStatistics": null,
                            "serviceEndpointStatistics": null
                        },
```

**Changes:**
- Adding customized status code for client generated 500

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 1]
```diff
@@ -448,6 +448,17 @@ public static class SubStatusCodes {
         public static final int NO_VALID_STORE_RESPONSE = 21009;
         public static final int SERVER_GENERATED_408 = 21010;
 
+        // SubStatusCodes for Client generated 500
+        public static final int MISSING_PARTITION_KEY_RANGE_ID_IN_CONTEXT = 21011;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:452]: Please also register these substatuscodes - and validate uniqueness in msdata repository. We decided to keep msdata as "ultimate truth" for client-side substatus codes

Please also add them to statusCodes.md in java github (shared with custoemrs and css)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Exceptions.java [hunk 2]
```diff
@@ -54,4 +54,8 @@ public static boolean isPartitionKeyMismatchException(CosmosException cosmosExce
         return Exceptions.isStatusCode(cosmosException, HttpConstants.StatusCodes.NOTFOUND) &&
             Exceptions.isSubStatusCode(cosmosException, HttpConstants.SubStatusCodes.PARTITION_KEY_MISMATCH);
     }
+
+    public static String getInternalServerErrorMessage(String prefix) {
+        return prefix + "-" + RMResources.InternalServerError;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Exceptions.java:59]:
```suggestion
        return prefix + " - " + RMResources.InternalServerError;
```

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
LGTM - except for making sure to "register" the sub staus codes in msdata and statusCodes.md

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Exceptions.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Exceptions.java]: updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java]: |500|               0 |Unknown| `Internal Server error` - Error returned from server, Indicates unexpected and unqualified internal service error.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|500|   20902 - 20910 |Unknown| `Internal Server error` - Client generated 500. The error message will have the details about the cause.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks!

**Pull request merged**